readme.md
# ScanPC
[![License](https://img.shields.io/badge/licence-AGPLv3-blue.svg)](https://www.gnu.org/licenses/agpl-3.0.en.html)
![Platform](https://img.shields.io/badge/platform-win--32%20%7C%20win--64-lightgrey.svg)
[![Language](https://img.shields.io/badge/language-python3-orange.svg)](https://www.python.org/)

ScanPC is an audit tool for Windows.  
It scans Windows computers to gather informations like :  
- the user accounts list
- the password policy
- the share folders list
- the hardware configuration
- the OS version
- the network interfaces
- the Windows updates (KB) list
- the softwares installed
- the firewall state
- the processes list
- the services list
- the antivirus state  

Started since October 2016  
Python 3.6 32 bit is used for all new developments.  
Project ongoing.  

## Compatibility
This software has been succesfully tested on the following Microsoft Windows systems :  
- XP x86
- 7 x86 and x64
- 10 x86 and x64

## Getting Started
**Download the project on your computer.**
```
git clone https://github.com/alevoski/ScanPC.git
```

**Install the required pip modules**
```
pip install -r requirements.txt
```

If you want to run it on Python 3.4 (for Windows XP compatibility) :
- you must install [pywin32-219.win32-py3.4.exe](https://sourceforge.net/projects/pywin32/files/pywin32/Build%20219/pywin32-219.win32-py3.4.exe/download)
- you must install psutil v3.4.2
```
pip install psutil==3.4.2
```
- you must install wmi (tested with v1.4.9)
```
pip install wmi
```

**Optional tool**  
RemoveDrive.exe from <https://www.uwe-sieber.de/drivetools_e.html> website is used to dismount device used to scan the computers.  


## HOW TO USE ?
**Method 1 - With Python and all dependencies**  
```
cd CS/
python main.py
```

**Method 2 - Compile the code into an executable file**  
You will need pyinstaller module
```
pip install pyinstaller
c:\<yourpythonpath>\Scripts\pyinstaller.exe --hiddenimport win32timezone --onefile --icon=logoScanPC.ico main.py
```
A folder named "dist" will be created with a scanPC executable file.  
***Note*** : you need to compile with a 32 bit Python version to perform scan on x86 and x64 Windows OS  
***Note 2*** : you need to compile with Python 3.4 x86 for a maximum OS compatibility  
***Note 3*** : you need [Visual C++ 2010 X86](https://www.microsoft.com/en-US/download/confirmation.aspx?id=5555) to execute a compiled Python program on the targeted computers

Put the code or the exe on a USB key and go scanning some Windows computers.  
Whatever method you chose, you will need **software_list.json** and **style.css** files in the same directory where you run ScanPC !  

***Note : To limit viruses spreading, you should always analyze your scanning devices between two scans !***  
[Decontamine_Linux can help you !](https://github.com/alevoski/decontamine_Linux)

#### Demo  
![](DOCS/DEMO/scanpc_demo.gif)  

## Author
Alexandre Buissé

## License
ScanPC. Audit tool for Windows.  
Copyright (C) 2019 Alexandre Buissé alevoski@pm.me

This program is free software: you can redistribute it and/or modify  
it under the terms of the GNU Affero General Public License as published  
by the Free Software Foundation, either version 3 of the License, or  
(at your option) any later version.  

This program is distributed in the hope that it will be useful,  
but WITHOUT ANY WARRANTY; without even the implied warranty of  
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the  
GNU Affero General Public License for more details.  

You should have received a copy of the GNU Affero General Public License  
along with this program.  If not, see <https://www.gnu.org/licenses/>.


DOCS/changelog.txt
CHANGELOG
Note : this project started in October 2016.
Many improvements have been done since this date until August 2018.
Then the project has been refreshed in March 2019 with major modifications, in a more pythonic way.

07/04/2020
    Changes in scan.py
        correction in process_info() to have full list of running processes with parent id and start date
        add persistence_info() with search_reg() and get_files() to find persistent objects in regedit and start folders

15/02/2020
    Change in readme.md
        Add parameter --hiddenimport win32timezone in pyinstaller compilation instruction
    New demo GIF

07/02/2020 - 09/02/2020
    Modules static analysis (# pylint3 --disable=line-too-long) + code cleaning and minor corrections
        main.py             9.71/10
        scans.py            9.69/10
        writer.py           10/10
        verif.py            9.61/10
        software_scan.py    9.73/10
        complement.py       9.72/10
    Mod security_product_info() from scans.py
        Same code had been identify 3 times : put in a new function security_product_get()
    Rename modules, functions, variables, etc. according PEP8 convention
    Merge av_date.py with complement.py
    Merge ask_dismount.py with verif.py
    Remove _rowheader() from writer.py
        Headers will be write with _row2tr
    Rename write() to writer()

27/07/2019
    Legacy update.
    scans.py module updated to work on Windows XP x86
    Functions patched:
        servicesInfo()
        securityProductInfo()
        systemInfo()

20/05/2019 - 24/05/2019
    software_scan module will try to find version number in software name if no version has been found in regedit
    complement module now shows results in an HTML table (thanks to a CSV file)
    Two new functions in scans module
        securityProductInfo() and securityProductState() which find informations about security products
            Antivirus
            FireWall
            AntiSpyware
    Improve visibility of the titles in the HTML report (thanks to <h2> tags)

13/05/2019 - 18/05/2019
    CSS is now define in style.css file
    software_scan module now shows if the software are up to date with sotware_list.json
        ok in green if up to date
        ko in red if not
    hotFixesInfo() now only lists security updates
    systemInfo() now shows more information about network interfaces (DNS, Default Gateway, Adapter enabled, ...)

06/05/2019 - 12/05/2019
    Log is now generated in HTML along multiple CSV files resulting of the various scans
    Many scans are now displayed in an HTML table
    In writer.py module
        4 news functions to manage CSV and HTML
            writeCSV()
            _row2tr()
            _rowheader()
            csv2html()
    In scans.py and software_scan.py modules
        Major modifications to generate CSV
    In main.py module
        readandcopy()
            It will now ask to open the log file in a web browser instead of notepad
            System call removed
        The files are generated in an unique folder

24/04/2019
    In software_scan.py module
        softwareInit()
            English text
            Total number of software found showned in log
        searchSoftware()
            Change conditions to keep a software
    In scans.py module
        calcFlag()
            Change condition to get flagTosave and min (otherwise it could be badly calculated)
    In complement.py module
        init()
            Change to get LAPS state (otherwise it is not detected)
    In av_date.py module
        English text

22/04/2019
    In software_scan.py module
        Code correction in softwareInit() to improve software table display
        Code correction in searchSoftware() to get software without version
    In scans.py module
        Manage an exception in hotFixesInfo() when a date cannot be parsed or does not exist

21/04/2019
    In writer.py module
        Clean code
            Functions removed
                writeFinalLog
                copyFile
        English doctrings
            prepaLogScan
            write
            writeLog
    In scan.py module
        English doctrings
            detectOS
    In av_date.py module
        Clean code
            Function removed
                initMcAfee
    In complement.py module
        Clean code
            Removed 2 global vars "username" and "computername"
            Replace "elemInLog" function name by "elemInList"
    In scans.py module
        Clean code
            Removed socket import
        update doctrings
        hotFixesInfo()
            Change KB date format : m/d/Y to Y-m-d
        servicesInfo(), processInfo() & portsInfo()
            Improve display
            Now return the list of elements found
            Show total number of elements found before listing
    In verif.py module
        English doctrings
    Better display of the final log
    Scanning time show at the end of the log

20/03/2019 to 20/04/2019
    First commit
    Major modifications 
        Many system calls replace by Python modules : only 3 left !
        Functions with French and English doctrings
        One unique log file is now generated during the all process
        The scan take less much time than before

CS/main.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
#ScanPC - main.py
#@Alexandre Buissé - 2019/2020

#Standard imports
import os
import time
import webbrowser
from datetime import datetime

#Project modules imports
import writer
import verif
import scans
from software_scan import software_init
from complement import complement_init

COMPUTERNAME = os.environ['COMPUTERNAME']
STYLECSSFILE = os.getcwd() + '/style.css'

def scanpart():
    '''
    **FR**
    Retourne la lettre du lecteur où est exécuté l'app
    **EN**
    Return the drive letter where the app is launched
    '''
    key = os.getcwd() # On prend le chemin du répertoire du script/exe
    key = str(key[:2])
    return key

def scanpc(log_file_path):
    '''
    **FR**
    Exécute le scan de l'ordinateur en appelant diverses fonctions,
    écrit les logs sur le périphérique qui a lancé l'exécutable,
    Retourne le chemin du log final
    **EN**
    Process the computer scan with several other functions,
    write logs on the device where the app has been launched,
    Return the final log path
    '''
    texte = "Computer scanning in progress... Please wait...\n"
    writer.writer(texte)

    log_file = str(log_file_path) + "FINAL.html"

    # Write first html tags
    elementraw = """<!DOCTYPE html>
    <html>
        <head>
            <title>ScanPC result</title>
            <meta charset="utf-8">
            <link rel="stylesheet" type="text/css" href="style.css">
        </head>
        <body>
            <p>
            {0}
            {1} <br>
            </p>"""
            # </body>
        # </html>"""

    # Texte d'intro
    texte = '<h1>Scanning of computer "' + COMPUTERNAME + '"</h1>'
    element2 = '<time>' + datetime.now().strftime("%A %d %B %Y %H:%M:%S") + '</time>\n'
    element = elementraw.format(texte, element2)
    writer.writelog(log_file, element)

    # Scans de base
    scans.user_info(log_file_path)
    scans.shared_folders_info(log_file_path)
    scans.hotfixes_info(log_file_path)
    scans.system_info(log_file_path)
    scans.security_product_info(log_file_path)
    scans.process_info(log_file_path)
    services_running_dict = scans.services_info(log_file_path)
    scans.ports_info(log_file_path)
    scans.persistence_info(log_file_path)

    texte4 = "\nBasic scan ended.\n"
    writer.writer(texte4)

    # Scan des logiciels
    software_dict = software_init(log_file_path)

    return log_file, software_dict, services_running_dict

def readandcopy(log_file):
    '''
    **FR**
    Proposer à l'utilisateur d'ouvrir le log passé en paramètre
    et de le copier sur l'ordinateur scanné
    **EN**
    Ask user if he wants to read the log
    and if he wants a copy on the scanned computer
    '''
    msg0 = "\nDo you want to read the scan report ? (y = yes, n = no)\n"
    writer.writer(msg0)
    if verif.reponse() == 'y':
        openers = webbrowser._tryorder
        opened = 0
        for browsers in openers:
            # print(browsers)
            if 'firefox' in browsers.lower() or 'iexplore' in browsers.lower():
                # print('Opening with ' + browsers)
                browser = webbrowser.get(webbrowser._tryorder[openers.index(browsers)])
                browser.open('file:' + str(log_file))
                opened = 1
                break
        if opened == 0:
            webbrowser.open(str(log_file))

    # Copie de fichier vers un emplacement défini
    msg0 = "\nDo you want a copy the scan report on the computer (C:\\ drive) ? (y = yes, n = no)\n"
    writer.writer(msg0)
    if verif.reponse() == 'y':
        # Copy log
        file_name = os.path.basename(log_file)
        file_to_write = "C:/" + file_name
        writer.copy_file(log_file, file_to_write)
        # Copy CSS file
        if os.path.isfile(STYLECSSFILE):
            file_name = os.path.basename(STYLECSSFILE)
            file_to_write = "C:/" + file_name
            writer.copy_file(STYLECSSFILE, file_to_write)

def fin(key):
    '''
    **FR**
    Si le périphérique est amovible, proposer à l'utilisateur de le retirer proprement
    **EN**
    If the device is removable, ask the user to safely remove (dismount) his device
    '''
    # Test si lecteur amovible
    texte = "Computer scanning ended, details have been saved on " + str(key) + ".\n"
    writer.writer(texte)
    if verif.amovible(key):
        # Proposer la déconnexion de la clé
        msg0 = "\nDo you want to dismount your device and quit ? (y = yes, n = no)\n"
        writer.writer(msg0)
        if verif.reponse() == 'y':
            verif.dismount(key)
    verif.quitter()

def init():
    '''
    **FR**
    Initialisation du programme
    **EN**
    Software initialization
    '''
    print('\t\t\t---------- ScanPC ---------')
    print('\t***************************************************************')
    print('\t*\t\t    Computer scanning                         *')
    print('\t*\t\t          Disclaimer !                        *')
    print('\t*\t This sofware only scan your computer                 *')
    print('\t*\t for users, various system information and software.  *')
    print('\t*\t No copy of your data will be performed.              *')
    print('\t***************************************************************')
    print()

    # Begin timer
    begin_scan = time.time()

    # Début du programme
    key = scanpart()
    unique_dir = str(COMPUTERNAME) + '_' + datetime.now().strftime('%d%m%y%H%M%S')
    log_file_path = str(key) + "/logScanPC/" + datetime.now().strftime('%Y/%m/%d/') + unique_dir + '/' + unique_dir
    # Appel de scanPC
    log_file, software_dict, services_running_dict = scanpc(log_file_path)
    # Tests complémentaires
    complement_init(log_file_path, software_dict, services_running_dict)
    writer.writelog(log_file, '\n</div>\n')

    # Ending timer
    end_scan = time.time()
    total_time_scan = end_scan - begin_scan
    elem = "----------------------- Scan ended ------------------------"
    writer.prepa_log_scan(log_file, elem)
    total_time = 'Computer analyzed in {} seconds.'.format(round(total_time_scan, 2))
    writer.writelog(log_file, total_time)
    writer.writer('Computer analyzed in ' + str(round(total_time_scan, 2)) + ' seconds.\n')

    # Write last html tags
    elementraw = """
        </body>
    </html>"""
    writer.writelog(log_file, elementraw)

    # Copy css file in the log directory
    if os.path.isfile(STYLECSSFILE):
        file_name = os.path.basename(STYLECSSFILE)
        file_to_write = str(key) + "/logScanPC/" + datetime.now().strftime('%Y/%m/%d/') + unique_dir + '/'  + file_name
        writer.copy_file(STYLECSSFILE, file_to_write)

    #Fin du programme
    readandcopy(log_file)
    fin(key)

if __name__ == '__main__':
    init()


CS/verif.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
#ScanPC - verif.py
#@Alexandre Buissé - 2019/2020

#Standard imports
import msvcrt
import os
import sys
import subprocess
import win32file

#Project modules imports
from writer import writer

def quitter():
    '''
    **FR**
    Quitter
    **EN**
    Quit
    '''
    msg = "Press any key to quit.\n"
    writer(msg)
    msvcrt.getch()
    os.system('cls')
    sys.exit(0)

def reponse():
    '''
    **FR**
    Réponse
    **EN**
    Response
    '''
    while True:
        rep = msvcrt.getch()
        if 'y' in str(rep):
            # print("machin ok")
            return 'y'
        elif 'n' in str(rep):
            return 'n'

def dismount(key):
    '''
    **FR**
    Démonter le périphérique passé en paramètre
    **EN**
    Dismount device passed in parameter
    '''
    my_app = "RemoveDrive.exe"
    remover, _ = verif_prog(my_app, key)
    if remover != '':
        params = key + ' -L -b' #-L =>boucle tant que la clé n'a pa pu être éjectée et -b pour afficher le msg Windows
        cmdline_user = my_app + ' ' + params
        #Obtenir le chemin du dossier du programme removedrive
        directory = os.path.dirname(os.path.abspath(remover))
        # print("test : ",cmdline_user)
        # print("test directory : ",directory)
        subprocess.call(cmdline_user, cwd=directory, shell=True, stdout=subprocess.PIPE, stdin=subprocess.PIPE, stderr=subprocess.PIPE)
    else:
        quitter()

def verif_prog(prog, key):
    '''
    **FR**
    Recherche d'un programme
    **EN**
    Search an app
    '''
    prog_find = ''
    prog_find_path = ''
    compteur = -1

    # Recherche
    prog_find, prog_find_path, compteur = find_that(prog, key)

    if compteur == -1:
        print("\nThis app needs " + str(prog) + " !")

    return prog_find, prog_find_path

def find_that(prog, folder):
    '''
    **FR**
    Recherche récursive
    **EN**
    Recursive search
    '''
    prog_find = ""
    prog_find_path = ""
    compteur = -1
    # print("Recherche dans => "+str(folder))
    folder = folder+"/"
    for root, _, filenames in os.walk(folder):
        for filename in filenames:
            # print("recherche de "+str(prog)+" dans "+str(folder))#"ok
            # print("filename :",filename)#ok
            if prog == filename:
                prog_find_path = os.path.abspath(root)
                # print("path prog :"+str(prog_find_path)) #test ok
                prog_find = os.path.abspath(root)+"/"+prog
                # root+"/"+prog
                # print (prog_find+" trouvé !\n") #ok
                compteur = 1
                return prog_find, prog_find_path, compteur
    return prog_find, prog_find_path, compteur

def amovible(key):
    '''
    **FR**
    Test si le périphérique passé en paramètre est amovible
    **EN**
    Test if device in parameter is removable
    '''
    return win32file.GetDriveType(key) == win32file.DRIVE_REMOVABLE


CS/writer.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
#ScanPC - writer.py
#@Alexandre Buissé - 2019/2020

'''
Writing module : transforms scan results into report (CSV + HTML)
'''

#Standard imports
from sys import stdout
from time import sleep
import os
from csv import DictWriter
from shutil import copy2

def copy_file(source, destination):
    '''
    Copy a file (source) to a destination
    '''
    file_name = os.path.basename(source)
    try:
        copy2(source, destination)
        if os.path.isfile(destination):
            msg = "\n" + file_name + " has been copied !\n"
            writer(msg)
        else:
            raise IOError
    except (IOError, PermissionError):
        msg = "\nUnable to copy " + file_name + " ! :\nPermission denied !"
        writer(msg)

def prepa_log_scan(logtowrite, elem):
    '''
    **FR**
    Ecrire un elément dans un fichier passés en paramètre
    **EN**
    Write elem in a file pass in parameter
    '''
    elem0 = '*' * len(elem)
    writelog(logtowrite, elem0 + '<br>\n' + elem + '<br>\n' + elem0 + '<br>\n<br>\n')

def writer(txt, timeout=100, eol=True):
    '''
    **FR**
    Forcer un affichage harmonieux (car sinon tout est affiché trop vite sans)
    **EN**
    Force a smooth display (because everything is printed to quickly without it)
    '''
    try:
        timeout = timeout / 400000  # ne pas faire le calcul à chaque itération !
        for char in txt:
            print(char, end='')
            stdout.flush()       # force l'affichage du buffer
            if not char in ' \n\t': # vérifie que char n'est pas dans la chaine " \n\t"
                sleep(timeout)
        if eol:
            print()
    except OSError:
        pass

def writelog(log_file, element):
    '''
    **FR**
    Enregistrement des logs
    **EN**
    Write logs
    '''
    # Créer un répertoire où seront repertorié tous les logs
    # print('log_file sent to writelog : ', log_file)
    dir2create = os.path.dirname(os.path.abspath(log_file))
    # print (dir2create)
    if not os.path.exists(dir2create):
        os.makedirs(dir2create)
    with open(log_file, mode='a', encoding='utf-8') as logfile:
        logfile.write(element)

def write_csv(log_file, fieldnames, element_dict):
    '''
    **FR**
    Enregistrement des logs au format CSV
    **EN**
    Write logs in CSV format
    '''
    with open(log_file, mode='w', newline='', encoding='utf-8-sig') as csvfile:
        writercsv = DictWriter(csvfile, delimiter=';', fieldnames=fieldnames)
        writercsv.writeheader()
        # for _, values in sorted(element_dict.items()):
        for _, values in element_dict.items():
            writercsv.writerow(values)

def _row2tr(row, attr):
    '''
    **FR**
    Transforme les lignes (issues d'un fichier CSV) en lignes de tableau HTML
    **EN**
    Transform rows (from a CSV file) in HTML table rows
    '''
    cols = row.split(';')
    bal1 = '<' + attr + '>'
    bal2 = '</' + attr + '>'
    tojoin = '<TR>'
    for data in cols:
        tojoin += bal1 + data + bal2
    tojoin += '</TR>'
    return tojoin

def csv2html(csv_file, summary):
    '''
    **FR**
    Transforme un fichier CSV en tableau HTML
    **EN**
    Transform a CSV file in an HTML table
    '''
    htmltxt = '<TABLE summary="' + summary + '">\n'
    with open(csv_file, mode='r', encoding='utf-8-sig') as csvfile:
        readercsv = csvfile.read()
    for rownum, row in enumerate(readercsv.split('\n')):
        if row != '':
            # Prepare header
            if rownum == 0:
                # print(row)
                htmlrow = _row2tr(row, 'TH')
            else:
                htmlrow = _row2tr(row, 'TD')
            htmlrow = '  <TBODY>%s</TBODY>\n' % htmlrow
            htmltxt += htmlrow
    htmltxt += '</TABLE>\n'
    return htmltxt


CS/av_date.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
#ScanPC - av_date.py
#@Alexandre Buissé - 2019

import winreg

def reg(HIVE, key, value):
    '''
    **FR**
    Interroge le registre Windows en fonction de la ruche, la clé et du nom de la valeur à rechercher
    dans la clé passés en paramètre
    Retourne la valeur de la clé
    **EN**
    Ask regedit with the hive, the key and the value to search, return the value of the key
    '''
    registry_key = winreg.OpenKey(HIVE, key, 0, winreg.KEY_READ)
    value, regtype = winreg.QueryValueEx(registry_key, value)
    winreg.CloseKey(registry_key)
    return value
    
def getMcAfee():
    '''
    **FR**
    Interroge le registre Windows à propos des informations qu'il a sur McAfee
    **EN**
    Ask regedit for McAfee informations
    '''
    hive = winreg.HKEY_LOCAL_MACHINE
    
    key = r"SOFTWARE\Wow6432Node\Network Associates\ePolicy Orchestrator\Application Plugins\VIRUSCAN8800"
    version = "DATVersion"
    date = "DatDate"
    version = reg(hive,key,version).split('.')[0]
    date = reg(hive,key,date)
    date = (date[0:4],date[4:6],date[6:8])
    
    keySrvLst = r"SOFTWARE\WOW6432Node\Network Associates\ePolicy Orchestrator\Agent"
    epoLst = "ePOServerList"
    srvList = reg(hive, keySrvLst, epoLst)
    lstmsg = "`\n--EPO servers list--<br>"
    # print("{d[2]}/{d[1]}/{d[0]} ({v})".format(d=date,v=version))
    return "Date & version : {d[2]}/{d[1]}/{d[0]} ({v})".format(d=date,v=version), lstmsg + srvList.replace(';', '\n')

def getWsus():
    '''
    **FR**
    Interroge le registre Windows à propos des informations qu'il a sur WSUS
    **EN**
    Ask regedit for WSUS informations
    '''
    hive = winreg.HKEY_LOCAL_MACHINE
    keySrv = r"SOFTWARE\Policies\Microsoft\Windows\WindowsUpdate"
    srv = "WUServer"
    srvWSUS = reg(hive, keySrv, srv)
    return srvWSUS

if __name__ == '__main__':
    date, epoLst = getMcAfee()
    print(date)
    print(epoLst)
    srvWSUS = getWsus()
    print(srvWSUS)

CS/software_scan.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
#ScanPC - software_scan.py
#@Alexandre Buissé - 2019/2020

'''
Scanning module : searches installed software in registry
'''

#Standard imports
import winreg
import os
import json
import re
from unicodedata import normalize

#Project modules imports
import writer

HIVE = winreg.HKEY_LOCAL_MACHINE
COMPUTERNAME = os.environ['COMPUTERNAME']
OTHER_SOFT_LIST = ['Java Auto Updater']

def reg(thekey, value):
    '''
    **FR**
    Interroge le registre Windows en fonction de la ruche, la clé et du nom de la valeur à rechercher
    dans la clé passés en paramètre (32 et 64 bits)
    Retourne la valeur de la clé
    **EN**
    Ask regedit with the HIVE, the key and the value to search (32 and 64 bits)
    Return the value of the key
    '''
    registry_key = None
    try:
        registry_key = winreg.OpenKey(HIVE, thekey, 0, winreg.KEY_READ)
    except FileNotFoundError:
        try:
            registry_key = winreg.OpenKey(HIVE, thekey, 0, winreg.KEY_READ | winreg.KEY_WOW64_64KEY)
        except FileNotFoundError:
            try:
                registry_key = winreg.OpenKey(HIVE, thekey, 0, winreg.KEY_READ | winreg.KEY_WOW64_32KEY)
            except FileNotFoundError:
                pass
    # print(registry_key)
    if registry_key is not None:
        try:
            value, _ = winreg.QueryValueEx(registry_key, value)
        except Exception:
            value = None
        winreg.CloseKey(registry_key)
    return value

def software_init(log_file_path):
    '''
    **FR**
    Initialisation et fonction principale de la recherche des logiciels
    **EN**
    Init the main function to search installed software
    '''
    texte5 = "Software scanning in progress...\n"
    writer.writer(texte5)

    # 1 - Début de l'analyse des logiciels
    log_file = log_file_path + "final.html"
    writer.writelog(log_file, '<div><br>\n')
    elem = '<h2>Software informations of computer "' + COMPUTERNAME + '"</h2>'
    writer.prepa_log_scan(log_file, elem)

    # 2 - Obtenir la liste des logiels de l'ordinateur
    regkey = r"SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall" #32 bits
    regkey2 = r"SOFTWARE\Wow6432Node\Microsoft\Windows\CurrentVersion\Uninstall" #64 bits
    software_list1 = software_lst(regkey)
    software_dict1 = search_software(software_list1, regkey)
    software_list2 = software_lst(regkey2)
    software_dict2 = search_software(software_list2, regkey2)

    # 3 - Concaténation des deux dictionnaires de valeurs des clés de registre (32b et 64b)
    final_dict = software_dict1
    for item, value in software_dict2.items():
        # print(item, value)
        if item not in final_dict.items():
            # print(item)
            final_dict.update({item:value})

    # 4 - Ecriture du fichier CSV
    header = ['Name', 'Version', 'Last Update', 'Last Version', 'Up to date', 'Publisher', 'Location']
    csv_file = log_file_path + "software.csv"
    writer.write_csv(csv_file, header, final_dict)

    # 5 - Transformation du CSV en HTML
    htmltxt = writer.csv2html(csv_file, 'Installed software')

    # 6 - Placer des id sur certains tags du tableau
    htmltxt = htmltxt.replace('<TD>ko</TD>', '<TD id="ko">ko</TD>')
    htmltxt = htmltxt.replace('<TD>ok</TD>', '<TD id="ok">ok</TD>')

    # 7 - Bilan
    soft_ok = 0
    soft_ko = 0
    others = 0
    for _, value in final_dict.items():
        stat = value['Up to date']
        if stat == 'ok':
            soft_ok += 1
        elif stat == 'ko':
            soft_ko += 1
        else:
            others += 1

    soft_summary_dict = {}
    soft_summary_dict[str(len(final_dict))] = {'Total Software':str(len(final_dict)),
                                               'Up to date':str(soft_ok), '% up to date':str(round(soft_ok/len(final_dict)*100)),
                                               'Outdated':str(soft_ko), '% outdated':str(round(soft_ko/len(final_dict)*100)),
                                               'others':str(others), '% others':str(round(others/len(final_dict)*100))}

    header = ['Total Software', 'Up to date', '% up to date', 'Outdated', '% outdated', 'others', '% others']
    csv_file = log_file_path + "software_summary.csv"
    writer.write_csv(csv_file, header, soft_summary_dict)
    htmltxt2 = writer.csv2html(csv_file, 'Software summary')
    htmltxt2 = htmltxt2.replace('<TH>Up to date</TH>', '<TH id="ok">Up to date</TH>')
    htmltxt2 = htmltxt2.replace('<TH>% up to date</TH>', '<TH id="ok">% up to date</TH>')
    htmltxt2 = htmltxt2.replace('<TH>Outdated</TH>', '<TH id="ko">Outdated</TH>')
    htmltxt2 = htmltxt2.replace('<TH>% outdated</TH>', '<TH id="ko">% outdated</TH>')

    # 8 - Ecriture fin du log
    writer.writelog(log_file, str(len(final_dict)) + ' software found :\n')
    writer.writelog(log_file, htmltxt)
    writer.writelog(log_file, '<br>Software summary :')
    writer.writelog(log_file, htmltxt2)
    writer.writelog(log_file, '\n</div>\n')
    texte5bis = "Software scanning ended.\n"
    writer.writer(texte5bis)

    return final_dict

def software_lst(thekey):
    '''
    **FR**
    Utiliser le registre de Windows pour obtenir les logiciels installés
    Retourne la liste des logiciels installés
    **EN**
    Use regedit to get all installed software
    Return a list of them
    '''
    registry_key = None
    software_list = None
    try:
        # registry_key = winreg.OpenKey(HIVE, thekey, 0, winreg.KEY_READ)
        registry_key = winreg.OpenKey(HIVE, thekey, 0, winreg.KEY_READ | winreg.KEY_WOW64_64KEY)
    except FileNotFoundError:
        try:
            registry_key = winreg.OpenKey(HIVE, thekey, 0, winreg.KEY_READ | winreg.KEY_WOW64_32KEY)
        except FileNotFoundError:
            try:
                registry_key = winreg.OpenKey(HIVE, thekey, 0, winreg.KEY_READ)
            except FileNotFoundError:
                pass
    # print(registry_key)
    if registry_key is not None:
        i = 0
        software_list = []
        while True:
            try:
                software_list.append(winreg.EnumKey(registry_key, i)) # ok pour 1 clé
                i += 1
                # print("value :", str(registry_key))
            except Exception:
                # print(Exception)
                winreg.CloseKey(registry_key)
                break
        # print(software_list)
    return software_list

def is_up_to_date(name_soft, version_soft):
    '''
    Compare version of the installed software with the last version known in software_list.json
    Return 'ok' if the software is up to date or 'ko' if not
    Return also the last known update and the last known version
    '''
    # Voir si les logiciels sont à jour
    # Open json software list
    software_list_file = os.getcwd() + '/software_list.json'
    with open(software_list_file, mode='r', encoding='utf-8') as softfile:
        data = json.load(softfile)
    # Compare software version in the json list with the dict
    for soft in data['software']:
        stat = ''
        last_update = ''
        last_version = ''
        # Match software
        if soft['name'].lower() in name_soft[:len(soft['name'])].lower():
            stat = 'ko'
            last_update = soft['last_update']
            last_version = soft['last_version']
            # Match version
            if soft['last_version'] in version_soft:
                stat = 'ok'
            break
    return stat, last_update, last_version

def search_version_in_name(name_soft):
    '''
    Search and return a version number from a software name
    '''
    try:
        return str(re.findall(r"(\d{0,2}(\.\d{1,2}).*?)", name_soft)[0][0])
    except IndexError:
        return ''

def search_software(software, thekey):
    '''
    **FR**
    Utiliser la liste des logiciels installés pour interroger le registre de Windows
    Retourne le dictionnaire des logiciels installés contenant la version, l'éditeur
    et le chemin d'installation
    **EN**
    Use the list of installed software to ask regedit
    Return the dict of installed software with version, publisher and install location
    '''
    name = "DisplayName"
    version = "DisplayVersion"
    publisher = "Publisher"
    path = "InstallLocation"
    software_dict = {}
    try:
        for soft in software:
            softkey = thekey+'\\'+soft
            name_soft = reg(softkey, name)
            version_soft = reg(softkey, version)
            publisher_soft = reg(softkey, publisher)
            path_soft = reg(softkey, path)
            # Conditions : some software does not have version & avoid Microsoft Updates
            if name_soft is not None and name_soft != '' and 'Update for' not in name_soft and publisher_soft is not None:
                if normalize('NFKD', name_soft.title()) not in software_dict:
                    # Is version empty then look for it in name
                    if version_soft is None:
                        version_soft = search_version_in_name(name_soft)
                    # Is software up to date ?
                    stat = ''
                    last_update = ''
                    last_version = ''
                    if name_soft not in OTHER_SOFT_LIST: # Do nothing if software is in OTHER_SOFT_LIST
                        stat, last_update, last_version = is_up_to_date(name_soft, version_soft)
                    # Write in dict
                    software_dict[normalize('NFKD', name_soft.title())] = {'Name':normalize('NFKD', name_soft.title()),
                                                                           'Version':version_soft,
                                                                           'Last Update':last_update,
                                                                           'Last Version':last_version,
                                                                           'Up to date':stat,
                                                                           'Publisher':publisher_soft,
                                                                           'Location':path_soft}
    except TypeError:
        pass
    return software_dict
